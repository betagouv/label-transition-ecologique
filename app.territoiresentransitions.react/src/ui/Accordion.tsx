import DOMPurify from 'dompurify';
import {useToggle} from 'ui/shared/useToggle';

export type TAccordionProps = {
  /** classes css supplémentaires */
  className?: string;
  /** identifiant dans le dom */
  id: string;
  /** titre de l'accordéon */
  titre: string;
  /** contenu de l'accordéon */
  html: string;
  /** état initial (par défaut false = replié) */
  initialState?: boolean;
};

/** Affiche une description détaillée dans un composant accordéon  */
export const Accordion = (props: TAccordionProps) => {
  const {className, id, titre, html, initialState} = props;
  const [expanded, toggle] = useToggle(initialState || false);

  return (
    <section className={`fr-accordion ${className || ''}`}>
      <h3 className="fr-accordion__title">
        <button
          className="fr-accordion__btn font-normal"
          aria-controls={id}
          aria-expanded={expanded}
          onClick={() => toggle()}
        >
          <i className="fr-fi-information-fill text-[#0063cb] fr-mr-3v" />
          {titre}
        </button>
      </h3>
      <div
        className={`fr-collapse${expanded ? '--expanded' : ''}`}
        id={id}
        dangerouslySetInnerHTML={{__html: DOMPurify.sanitize(html)}}
      />
    </section>
  );
};
