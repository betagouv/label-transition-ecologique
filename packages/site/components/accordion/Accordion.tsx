'use client';

import classNames from 'classnames';
import {useEffect, useState} from 'react';
import {processMarkedContent} from 'src/utils/processMarkedContent';

type AccordionType = {
  id: string;
  title: string;
  content: string | React.ReactNode;
  initialState?: boolean;
};

const Accordion = ({
  id,
  title,
  content,
  initialState = false,
}: AccordionType) => {
  const [expanded, setExpanded] = useState(initialState ?? false);
  const [processedContent, setProcessedContent] = useState<
    string | undefined
  >();

  const processContent = async (content: string) => {
    const newContent = await processMarkedContent(content);
    setProcessedContent(newContent);
  };

  useEffect(() => {
    if (typeof content === 'string') processContent(content);
  }, [content]);

  const contentClassName = classNames({
    'fr-collapse--expanded py-3 px-4': expanded,
    'fr-collapse': !expanded,
  });

  return (
    <div className="fr-accordion">
      <h3 className="fr-accordion__title">
        <button
          className="fr-accordion__btn nav-btn"
          aria-controls={id}
          aria-expanded={expanded}
          onClick={() => setExpanded(prevState => !prevState)}
        >
          {title}
        </button>
      </h3>
      {processedContent ? (
        <div
          className={contentClassName}
          id={id}
          dangerouslySetInnerHTML={{
            __html: processedContent,
          }}
        />
      ) : (
        <div className={contentClassName} id={id}>
          {content}
        </div>
      )}
    </div>
  );
};

export default Accordion;
