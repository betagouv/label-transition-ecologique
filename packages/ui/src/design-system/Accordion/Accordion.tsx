import {useState} from 'react';
import classNames from 'classnames';
import {Icon, IconValue} from '@design-system/Icon';

type AccordionType = {
  id?: string;
  /** Icône ajoutée avant le titre */
  icon?: IconValue;
  /** Pour changer la position de l'icône */
  iconPosition?: 'left' | 'right';
  /** Titre */
  title: string;
  /** Contenu (chaîne ou composant React personnalisé) */
  content: string | React.ReactNode;
  /** Permet de styler le contenu (ignoré si le contenu n'est pas une chaîne) */
  contentClassname?: string;
  /** Pour afficher un texte optionnel à côté du titre */
  subtitle?: boolean;
};

/**
 * Affiche un contenu complémentaire
 */
export const AccordionControlled = ({
  icon,
  iconPosition = 'left',
  id,
  title,
  content,
  contentClassname,
  expanded,
  setExpanded,
  subtitle,
}: AccordionType & {
  /** le contenu est visible */
  expanded: boolean;
  /** appelée pour déplier/replier le contenu */
  setExpanded: (value: boolean) => void;
}) => {
  return (
    <>
      {/** EN-TËTE */}
      <div
        role="button"
        aria-controls={id}
        aria-expanded={expanded}
        onClick={() => setExpanded(!expanded)}
        className={classNames(
          'border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1',
          {'border-transparent rounded-b-none bg-primary-2': expanded}
        )}
      >
        {/** picto indiquant l'état ouvert/fermé */}
        <Icon
          icon="arrow-right-s-line"
          className={classNames('mr-3 text-primary transition-transform', {
            'transform: rotate-90': expanded,
          })}
        />
        {/** icône optionnelle avant le titre */}
        {!!icon && iconPosition === 'left' && (
          <Icon icon={icon} className="mr-3 text-primary" />
        )}
        {/** titre */}
        {title}
        {/** sous-titre */}
        {!!subtitle && (
          <span className="font-normal text-xs ml-3 line-clamp-1">
            {subtitle}
          </span>
        )}
        {/** icône optionnelle alignée à droite */}
        {!!icon && iconPosition === 'right' && (
          <Icon icon={icon} className="ml-auto text-primary" />
        )}
      </div>

      {/** CONTENU */}
      {!!expanded &&
        (typeof content === 'string' ? (
          <div
            className={classNames(
              'p-4 border border-t-0 border-grey-4 rounded-b-lg',
              contentClassname
            )}
            id={id}
          >
            {content}
          </div>
        ) : (
          content
        ))}
    </>
  );
};

export const Accordion = ({
  initialState,
  ...props
}: AccordionType & {initialState?: boolean}) => {
  const [expanded, setExpanded] = useState(initialState || false);

  return (
    <AccordionControlled
      {...props}
      expanded={expanded}
      setExpanded={setExpanded}
    />
  );
};
