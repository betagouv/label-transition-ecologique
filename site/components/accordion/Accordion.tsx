'use client';

import classNames from 'classnames';
import DOMPurify from 'dompurify';
import {marked} from 'marked';
import {useState} from 'react';

type AccordionType = {
  id: string;
  title: string;
  content: string | React.ReactNode;
  initialState?: boolean;
};

const Accordion = ({
  id,
  title,
  content,
  initialState = false,
}: AccordionType) => {
  const [expanded, setExpanded] = useState(initialState ?? false);

  const contentClassName = classNames({
    'fr-collapse--expanded py-3 px-4': expanded,
    'fr-collapse': !expanded,
  });

  return (
    <div className="fr-accordion">
      <h3 className="fr-accordion__title">
        <button
          className="fr-accordion__btn"
          aria-controls={id}
          aria-expanded={expanded}
          onClick={() => setExpanded(prevState => !prevState)}
        >
          {title}
        </button>
      </h3>
      {typeof content === 'string' ? (
        <div
          className={contentClassName}
          id={id}
          dangerouslySetInnerHTML={{
            __html: DOMPurify.sanitize(marked.parse(content)),
          }}
        />
      ) : (
        <div className={contentClassName} id={id}>
          {content}
        </div>
      )}
    </div>
  );
};

export default Accordion;
