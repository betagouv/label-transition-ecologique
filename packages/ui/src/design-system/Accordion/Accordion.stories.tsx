import { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Accordion, AccordionControlled } from './Accordion';
import DoubleCheckIcon from '@tet/ui/assets/DoubleCheckIcon';

const meta: Meta<typeof Accordion> = {
  component: Accordion,
  args: {
    title: 'Titre',
    content: 'contenu',
  },
};

export default meta;

type Story = StoryObj<typeof Accordion>;

export const Default: Story = {};

export const ControlleOuvert = () => (
  <AccordionControlled
    title="Titre"
    content="Contenu"
    expanded
    setExpanded={action('setExpanded')}
  />
);

export const ContenuStyle: Story = {
  args: {
    content: 'Contenu stylé',
    contentClassname: 'border border-dashed',
  },
};

export const ContenuPersonnalise: Story = {
  args: {
    content: <p className="font-bold border p-8">Contenu personnalisé</p>,
  },
};

export const AvecIcone: Story = {
  args: {
    icon: 'information-line',
  },
};

export const AvecIconeADroite: Story = {
  args: {
    icon: 'information-line',
    iconPosition: 'right',
  },
};

export const AvecIconeSVG: Story = {
  args: {
    icon: <DoubleCheckIcon className="fill-primary-7" />,
  },
};

export const AvecSousTitre: Story = {
  args: {
    icon: 'lock-line',
    subtitle:
      'Cette section est visible uniquement par les membres de votre collectivité',
  },
};

export const CompletTitreLong: Story = {
  args: {
    icon: 'lock-line',
    iconPosition: 'right',
    title:
      'Titre long qui prend plusieurs lignes, qui est vraiment très long et qui ne finit pas de finir',
    subtitle:
      'Cette section est visible uniquement par les membres de votre collectivité',
  },
};
