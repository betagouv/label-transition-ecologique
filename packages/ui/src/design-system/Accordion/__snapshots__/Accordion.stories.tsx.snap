// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Accordion AvecIcone smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  <span class="ri-information-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary">
  </span>
  Titre
</div>
`;

exports[`design-system/Accordion AvecIconeADroite smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  Titre
  <span class="ri-information-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] ml-auto text-primary">
  </span>
</div>
`;

exports[`design-system/Accordion AvecSousTitre smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  <span class="ri-lock-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary">
  </span>
  Titre
  <span class="font-normal text-xs ml-3 line-clamp-1">
    Cette section est visible uniquement par les membres de votre collectivit√©
  </span>
</div>
`;

exports[`design-system/Accordion ContenuPersonnalise smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  Titre
</div>
`;

exports[`design-system/Accordion ContenuStyle smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  Titre
</div>
`;

exports[`design-system/Accordion ControlleOuvert smoke-test 1`] = `
<div role="button"
     aria-expanded="true"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1 border-transparent rounded-b-none bg-primary-2"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform transform: rotate-90">
  </span>
  Titre
</div>
<div class="p-4 border border-t-0 border-grey-4 rounded-b-lg">
  Contenu
</div>
`;

exports[`design-system/Accordion Default smoke-test 1`] = `
<div role="button"
     aria-expanded="false"
     class="border border-grey-4 rounded-lg h-16 flex items-center px-8 font-bold text-primary-10  hover:bg-primary-1"
>
  <span class="ri-arrow-right-s-line font-normal text-center h-5 w-5 text-[1.25rem] leading-[1.25rem] mr-3 text-primary transition-transform">
  </span>
  Titre
</div>
`;
